### DISCUSSION TITLE
Why is v9 complaining about `import` in `eslint.config.mjs`?

### DISCUSSION THREAD

**Discussion by MetRonnie** [eslint/eslint#18300] [Q&A]
```js
/* eslint.config.mjs */
import js from "@eslint/js"
// ^^^ Parsing error: 'import' and 'export' may appear only with 'sourceType: module'

export default [
    js.configs.recommended,
    {
        languageOptions: {
            sourceType: "commonjs",
        },
    },
]
```

Sure I have set `sourceType: "commonjs"`, but why does that seem to preclude the possibility of using an `mjs` ESLint config? It still runs successfully.

`package.json`:
```json
{
    "name": "foo",
    "private": true,
    "devDependencies": {
        "@eslint/js": "^9.0.0",
        "eslint": "^9.0.0",
    }
}
```

**Comment by Rec0iL99** [comment]
Hi @MetRonnie, is this error coming from your editor? Could you please provide a reproducible example for us to try out? Thanks.

**Reply by MetRonnie** [reply]
This is reported by `npx eslint .`. I've added my `package.json`, the example should be reproduceable as is

**Reply by Rec0iL99** [reply]
I get it now. Sorry, I missed the `sourceType` point. I think since you explicitly put `sourceType: "commonjs"` for all files in your config, the parser expects every file it is parsing to be of `commonjs` format (`require()`) regardless of using `.mjs` file extension. 

https://github.com/eslint/eslint/issues/17863

cc @eslint/eslint-tsc

**Reply by Rec0iL99** [reply]
```js
/* eslint.config.mjs */
import js from "@eslint/js";

export default [
  js.configs.recommended,
  {
    languageOptions: {
      sourceType: "commonjs",
    },
  },
  {
    files: ["eslint.config.mjs"],
    languageOptions: {
      sourceType: "module",
    },
  },
];
```

You can do this as a way to tell the parser that your config is `sourceType: "module"`.

**Reply by voxpelli** [reply]
Is there a way to not have any specific `sourceType` set? I think that was how I used ESLint before version 9

**Reply by MetRonnie** [reply]
I would have thought the `sourceType` option would apply to `.js` files only, seeing as `.cjs` and `.mjs` are already explicit about what type they are (to me it doesn't ever make sense that an `.mjs` file would be treated as commonJS or that a `.cjs` file would get treated as ESM)

**Reply by nzakas** [reply]
`sourceType` is required for all files. By default it's set to `module`, which means the same thing as it does in `package.json` -- it's an ESM module.

By default, any files ending with `.mjs` are automatically assigned `sourceType: "module"` and `.cjs` files are automatically assigned `sourceType: "commonjs"`. The config you shared in your first post overrides `sourceType` for every type of file, including `.mjs`. `sourceType: "commonjs"` does not support `import` and `export`.

The important thing to understand is that CommonJS *is a different type of module* than ESM, so ESLint needs to parse and interpret it in a different way. Because you are using `.mjs`, there's no reason to set `sourceType: "commonjs"`.

See: https://eslint.org/docs/latest/use/configure/language-options#specifying-javascript-options

**Reply by MetRonnie** [reply]
My repo predominantly comprises `.js` files that are CommonJS which is why I set `sourceType: "commonjs"`. But I prefer ESM syntax and so would rather use `.mjs` for non-source files.

Although if I set `sourceType: "module"` ESLint runs ok on the CommonJS `.js` files, which I don't understand.

**Reply by nzakas** [reply]
If you're using a combination of ESM and CommonJS in your project, then it's best to configure changes targeted at just the files you want to have that configuration: https://eslint.org/docs/latest/use/configure/configuration-files#specifying-files-and-ignores

> Although if I set `sourceType: "module"` ESLint runs ok on the CommonJS `.js` files, which I don't understand.

ESM syntax is a superset of standard JavaScript syntax, which CommonJS is. That's why it works. There are differences in how ESLint handles these two file types, specifically around scoping, but for the most part this should work as long as you manually define globals like `require` and `exports`.

**Reply by yoyo837** [reply]
Sorry, I don't quite understand that I have some files `.mjs`, and the others are `.js` (Commonjs). Why can't it use ESM to parse the MJS file by default? I received an error `Parsing error: 'import' and 'export' may appear only with 'sourceType: module` in MJS file. https://github.com/eslint/eslint/issues/19004

**Reply by mdjermanovic** [reply]
> Why can't it use ESM to parse the MJS file by default?

`.mjs` files are parsed as ESM by default. If you set `sourceType: "commonjs"` in your config file, they will be parsed as CommonJS.

**Reply by voxpelli** [reply]
Since https://github.com/eslint/config-inspector/pull/66 the config-inspector includes the default config, so using the config-inspector one can debug which config it is that overrides the default config for a specific file
